lineapy/utils/config.py

```diff
@@ -1,40 +1,181 @@
+import json
 import logging
+import os
+from dataclasses import dataclass
+from os import PathLike
 from pathlib import Path
+from typing import Any, Optional
 
-FOLDER_NAME = ".lineapy"
-
-LOG_FILE_NAME = "linea.log"
-
+LINEAPY_FOLDER_NAME = ".lineapy"
+LOG_FILE_NAME = "lineapy.log"
+CONFIG_FILE_NAME = "lineapy_config.json"
+FILE_PICKLER_BASEDIR = "linea_pickles"
+DB_FILE_NAME = "db.sqlite"
 CUSTOM_ANNOTATIONS_FOLDER_NAME = "custom-annotations"
-
 CUSTOM_ANNOTATIONS_EXTENSION_NAME = ".annotations.yaml"
 
 logger = logging.getLogger(__name__)
 
 
-def linea_folder() -> Path:
-    """
-    Linea folder exists at the root user level (via `Path.home()`).
-    """
-    # TODO: add a special case for tests (or some scoping config)
-    linea_folder = Path.home() / FOLDER_NAME
-    if not linea_folder.exists():
-        logger.warning(
-            f"No {FOLDER_NAME} folder found. Creating a new folder in {Path.home()} directory."
-        )
-        linea_folder.mkdir(parents=False, exist_ok=True)
-    return linea_folder
+@dataclass
+class LineapyConfig:
+    """LineaPy Configuration
 
+    A dataclass that holds configuration items and sets them as environmental
+    variables. All items are initialized with default value. Then replace
+    with values in the configuration file (if it is existing in LINEAPY_HOME_DIR,
+    use this one, otherwise look for home directory) if available. Then,
+    replace with values in environmental variables if possible. Finally, it sets
+    all values in environmental variables.
 
-def custom_annotations_folder() -> Path:
-    """
-    Where annotations are stored within linea folder.
     """
-    parent = linea_folder()
-    annotate_folder = parent / CUSTOM_ANNOTATIONS_FOLDER_NAME
-    if not annotate_folder.exists():
-        logger.warning(
-            f"No {CUSTOM_ANNOTATIONS_FOLDER_NAME} folder found. Creating a new folder in {parent} directory."
-        )
-        annotate_folder.mkdir(parents=False, exist_ok=True)
-    return annotate_folder
+
+    home_dir: Path
+    database_connection_string: Optional[str]
+    # artifact_storage_backend: str
+    artifact_storage_dir: Optional[Path]
+    customized_annotation_folder: Optional[Path]
+    do_not_track: bool
+    logging_level: str
+    logging_file: Optional[PathLike]
+    # artifact_aws_s3_bucket : str
+    # artifact_aws_s3_bucket_prefix : str
```

Comment:
at some point these configs should be nested (to indicate that they should be not null together?)

Reply:
Plan to use fsspec, should be able to identify filesystem backend in `FILE_PICKLER_BASEDIR` directly.